Add support for lighting channels

hosea1008 opened the pull request:
## What does this PR do?

Our artist found that it is necessary to support shining only some objects by only some lights, they were asking for a `lighting channel` functionality in O3DE, like what unreal engine is supported. 

So I implemented this feature. Considering this feature do need to change the lighting so I have to implemented in the engine, we are happy to contribute this feature to the community.

Briefly, `lighting channel` introduces 3 channel for each kind of lights and lightable objects, only the light and the objects in the same channel can be lighted. The default behavior is all lights and objects in channel 0, so it should not change the current way to work with lighting, but it provide a further option to grant the artist the flexibility to control which light shines which objects.

I implemented it by introducing a channel mask flag in each light types, and the meshes' and actors' object srg, and in each light's apply lighting function, it would evaluate whether the light and the object are in the same channel first.

<img width="246" alt="lighting channel lights" src="https://github.com/o3de/o3de/assets/5073925/b28bd065-a1e2-4191-8bc5-d0cfc7f3c890">
<img width="249" alt="lighting channel mesh" src="https://github.com/o3de/o3de/assets/5073925/59cf08aa-84ab-4cad-84e6-a3ab8bd443ca">


This PR introduce the functionality of lighting channel support similar to unreal engine into O3DE.

The lighting channel in unreal engine works like this: https://docs.unrealengine.com/5.3/en-US/using-lighting-channels-in-unreal-engine/

## How was this PR tested?

Self tested, and this feature has already been using in quite a lot of our real projects.

hosea1008 commented:
I firstly set the lighting channel mask in the lighting data from object srg, it is because I don't want to put the object srg inside the `XXXLight.azsli`, I think it would be better for such a commonly used shader header to have fewer dependency of SRGs. In our practice, we may have something need to be lighted other than ordinary meshes (like GBuffers in deferred pipeline), they may not have object SRGs, in this way, we can reuse the `ApplyXXXLight` function instead of rewrite it again.

DIFG-github commented:
Unless it impacts performance, wouldn't having the lighting channels use a key value or int value be a better approach? Its great to see support for lighting channels, though having a key value for what channel you are on could allow many more than 3 channels to give artists the same control used in film production lighting.

Just an opinion of course, and I am not sure if having it a key value with nearly infinite light channels would make a performance impact if they are not in use.

hosea1008 commented:
> Unless it impacts performance, wouldn't having the lighting channels use a key value or int value be a better approach? Its great to see support for lighting channels, though having a key value for what channel you are on could allow many more than 3 channels to give artists the same control used in film production lighting.
> 
> Just an opinion of course, and I am not sure if having it a key value with nearly infinite light channels would make a performance impact if they are not in use.

Actually it can already support up to 32 channels with `uint` value, but I am thinking whether we need so many lighting channels... In our practice, 3 lighting channels is enough for our projects, and unreal support only 3 lighting channels too.

I suppose if we have too many lighting channels, a lot of lights and objects, the artist might be easily forget what lights in what channels, and may easily mess up the lightings.

gadams3 commented on Gems/EMotionFX/Code/Source/Integration/Editor/Components/EditorActorComponent.h:
```diff
@@ -191,6 +192,11 @@ namespace EMotionFX
             ActorAsset::ActorInstancePtr        m_actorInstance;            ///< Live actor instance.
             AZStd::unique_ptr<RenderActorInstance> m_renderActorInstance;
 
+            uint32_t m_lightingChannelMask = 1;
+            bool m_lightingChannel0 = true;
+            bool m_lightingChannel1 = false;
+            bool m_lightingChannel2 = false;
```
This seems like a very useful feature. The biggest issue that I see is the way that the data is being represented in all of the components and controllers. Ideally, you would only need one variable stored and reflected per class to represent the lighting channel mask. I'm assuming that it's separated into blocks like this so that it's easier to reflect and edit in the edit context. This will make it extremely challenging if you want to add more than three lighting channels in the future.

My recommendation would be to change all of the variables in component based classes to use a single AZStd::array<bool, N> or AZStd::fixed_vector<bool, N>. N is the maximum number of lighting channels and should be defined globally. Then you only have to reflect one variable. The edit context will only need one entry and it should automatically create a checkbox for every element. Set up the data element attributes so that the container cannot be modified or reordered. Also make sure that it starts at a collapsed state so that it doesn't take up a lot of UI real estate by default.

IIRC, those containers have a specialization when the type is bool to store one element per bit.

I also recommend using uint32_t consistently. I see alternating uses of uint32_t and uint.

hosea1008 replied:
Hi, I changed the code to gather the channels inside a class, so that we won't need to update everywhere once we want more lighting channels.

For using uint32_t consistently, this version use only uint32_t in C++ side, and use uint in the shader.

gadams3 commented on Gems/Atom/Feature/Common/Code/Include/Atom/Feature/LightingChannel/LightingChannelConfiguration.h:
```diff
@@ -0,0 +1,40 @@
+        public:
+            AZ_TYPE_INFO(LightingChannelConfiguration, "{7FFD6D01-BABE-FE35-612F-63A30925E5F7}");
+
+            AZStd::array<bool, LightingChannelsCount> m_lightingChannelFlags;
```
If there was a property editor for it, or someone wanted to create one, I would recommend using AZStd::bitset<N> so you did not need to manually update the mask.

gadams3 replied:
```suggestion
            AZStd::array<bool, LightingChannelsCount> m_lightingChannelFlags{true, false, false, false, false};
```

gadams3 commented on Gems/Atom/Feature/Common/Code/Include/Atom/Feature/LightingChannel/LightingChannelConfiguration.h:
```diff
@@ -0,0 +1,40 @@
+    {
+        const uint32_t LightingChannelsCount = 5;
+
+        class LightingChannelConfiguration
```
Should this just be a struct?

gadams3 commented on Gems/Atom/Feature/Common/Code/Include/Atom/Feature/LightingChannel/LightingChannelConfiguration.h:
```diff
@@ -0,0 +1,40 @@
+
+            static void Reflect(AZ::ReflectContext* context);
+
+            uint32_t GetLightingChannelMask() const { return m_mask; }
```
This could compute the value and return a mask instead of having UpdateLightingChannelMask and m_mask. This will guarantee that everything is always synchronized without a manual update  step.

gadams3 commented on Gems/Atom/Feature/Common/Code/Include/Atom/Feature/LightingChannel/LightingChannelConfiguration.h:
```diff
@@ -0,0 +1,40 @@
+            static void Reflect(AZ::ReflectContext* context);
+
+            uint32_t GetLightingChannelMask() const { return m_mask; }
+            void SetLightingChannelMask(const uint32_t mask) { m_mask = mask; }
```
This probably needs to update the lighting channels container whenever it's called. If taking the other recommendation it will also eliminate the need for the mask variable and the manual update step.

gadams3 commented on Gems/Atom/Feature/Common/Code/Include/Atom/Feature/LightingChannel/LightingChannelConfiguration.h:
```diff
@@ -0,0 +1,40 @@
+            uint32_t GetLightingChannelMask() const { return m_mask; }
+            void SetLightingChannelMask(const uint32_t mask) { m_mask = mask; }
+
+            void UpdateLightingChannelMask();
+        private:
+            uint32_t m_mask {0x01};
```
These can probably go if all of the conversion logic is moved into the get/set functions.

gadams3 commented on Gems/Atom/Feature/Common/Code/Source/LightingChannel/LightingChannelConfiguration.cpp:
```diff
@@ -0,0 +1,53 @@
+            }
+        }
+
+        void LightingChannelConfiguration::UpdateLightingChannelMask()
+        {
+            AZ::u32 index = 0;
+            m_mask = 0;
+            AZStd::for_each(m_lightingChannelFlags.begin(), m_lightingChannelFlags.end(), [this, &index](const bool value){
+                this->m_mask |= (static_cast<AZ::u32>(value) << (index++));
+            });
+        }
+    }
```
This might be more efficient and compiler optimized/unrolled if you use a traditional for loop because the number of lighting channels is constant and you need the index anyway.

gadams3 reviewed (commented):
Thank you for implementing this and making the changes. I had some other comments on the lighting channels class to help bulletproof it and keep the data between the array and mask from getting out of sync by eliminating the mask variable and the update function, relying on the get/set functions to do the conversion.

kh-huawei reviewed (approved):
Looks good to me

moudgils commented on Gems/AtomLyIntegration/EMotionFXAtom/Code/Source/AtomActorInstance.cpp:
```diff
@@ -815,6 +816,27 @@ namespace AZ::Render
+                if (lightingChannelMaskIndex.IsValid())
+                {
+                    objectSrg->SetConstant(lightingChannelMaskIndex, m_actorInstance->GetLightingChannelMask());
+                    m_meshFeatureProcessor->QueueObjectSrgForCompile(*m_meshHandle);
```
Do we need to queue the SRG compilation so many times in this method? Shouldn't one suffice at the end of the method?

gadams3 commented:
Looks like tests fail for this change. Please run them locally to resolve any issues and then will retrigger the build.

hosea1008 commented:
> Looks like tests fail for this change. Please run them locally to resolve any issues and then will retrigger the build.

Yes, I fixed it. This test case seems to create mesh component controller when the mesh feature processor is not ready, now I added protection to it.

hosea1008 commented:
@moudgils @gadams3 Hi, I fixed and now all checks passed, can you help merging this PR? Thanks
